mbe: Macro fragment fields

joshtriplett opened the pull request:
Initial support for macro fragment fields (marked as incomplete).

This adds the `:fn` and `:adt` fragment types, and adds a `name` field
supported by those types. This provides an initial demonstration and minimal
initial infrastructure for macro fragment fields.

A macro fragment field operation takes a `ParseNtResult` (e.g.
`ParseNtResult::Fn`) and returns a new `ParseNtResult` (e.g.
`ParseNtResult::Ident`).

joshtriplett committed:
mbe: Add fragment type `fn` to parse a function

joshtriplett committed:
mbe: Initial support for macro fragment fields

This only supports a `name` field of `fn` and `adt` fragments, for now,
but it adds the infrastructure for adding more fields.

joshtriplett committed:
mbe: Add error test for macro fragment fields

joshtriplett committed:
mbe: Add fragment type `adt` to parse an ADT (`struct`, `enum`, or `union`)

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18007765503/job/51232092392) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51232092392) [(plain)](https://github.com/rust-lang/rust/commit/f9c13c17436c41876065afffc7b6320c18e546d7/checks/51232092392/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt: checked 6397 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
tidy [features]: Found 1 features without a gate test.
tidy [features]: FAIL
Expected a gate test for the feature 'macro_fragments_more'.
Hint: create a failing test file named 'tests/ui/feature-gates/feature-gate-macro-fragments-more.rs',
      with its failures due to missing usage of `#![feature(macro_fragments_more)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-macro_fragments_more line to the test file.
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.2)
linting python files
---
linting javascript files and applying suggestions
Running eslint on rustdoc JS files
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: features
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/npm --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1561:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1281:29

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:03:12
  local time: Thu Sep 25 12:47:38 UTC 2025
  network time: Thu, 25 Sep 2025 12:47:38 GMT
##[error]Process completed with exit code 1.
```

</details>

joshtriplett committed:
mbe: Support `vis` fragment field for `fn`; add infra for recursive token collection

Add support for getting a `.vis` field from a `fn` type. This requires
adding support for collecting tokens when parsing the visibility of a
function. Add a new `ForceCollect::Recursive`, and plumb that through
the item parser enough to support visibility. Only use that when a macro
uses macro metavariable expressions, as we don't need the tokens
otherwise.

`ForceCollect::Recursive` will need further handling later, to capture
the tokens of any other fields needed, but that can be added alongside
any fields it supports.

bors commented:
:umbrella: The latest upstream changes (presumably #146636) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18100187968/job/51500955489) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51500955489) [(plain)](https://github.com/rust-lang/rust/commit/b3e8379c3cc032934371141f43b706df694e6bd3/checks/51500955489/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] toml test:false 0.208
error[E0061]: this method takes 2 arguments but 1 argument was supplied
    --> src/tools/rustfmt/src/parse/macros/lazy_static.rs:38:29
     |
  38 |         let vis = parse_or!(parse_visibility, rustc_parse::parser::FollowedByType::No);
     |                             ^^^^^^^^^^^^^^^^  --------------------------------------- argument #1 of type `ForceCollect` is missing
     |
note: method defined here
    --> /checkout/compiler/rustc_parse/src/parser/mod.rs:1485:12
     |
1485 |     pub fn parse_visibility(
     |            ^^^^^^^^^^^^^^^^
help: provide the argument
     |
  38 |         let vis = parse_or!(parse_visibility(/* ForceCollect */, rustc_parse::parser::FollowedByType::No), rustc_parse::parser::FollowedByType::No);
     |                                             +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0061`.
[RUSTC-TIMING] rustfmt_nightly test:false 2.349
error: could not compile `rustfmt-nightly` (lib) due to 1 previous error
```

</details>

joshtriplett committed:
mbe: Support `.vis` on `adt` fragments

joshtriplett committed:
mbe: Support `.vis` on items
